panel/pane.py

```diff
@@ -376,29 +385,69 @@ def _png(self):
             r = requests.request(url=self.object, method='GET')
             return r.content
 
-    def _pngshape(self, data):
-        """Calculate and return PNG width,height"""
-        import struct
-        w, h = struct.unpack('>LL', data[16:24])
-        return int(w), int(h)
+    def _imgshape(self, data):
+        """Calculate and return image width,height"""
+        raise NotImplementedError
     
     def _get_properties(self):
-        data = self._png()
+        data = self._img()
         b64 = base64.b64encode(data).decode("utf-8")
-        src = "data:image/png;base64,{b64}".format(b64=b64)
+        src = "data:image/"+self.imgtype+";base64,{b64}".format(b64=b64)
         html = "<img src='{src}'></img>".format(src=src)
         
-        p = super(PNG,self)._get_properties()
-        width, height = self._pngshape(data)
+        p = super(Image,self)._get_properties()
+        width, height = self._imgshape(data)
         if self.width  is None: p["width"]  = width
         if self.height is None: p["height"] = height
         p["text"]=html
         return p
 
 
+class PNG(Image):
+
+    imgtype = 'png'
+    
+    def _imgshape(self, data):
+        import struct
+        w, h = struct.unpack('>LL', data[16:24])
+        return int(w), int(h)
+
+
+class GIF(Image):
+
+    imgtype = 'gif'
+    
+    def _imgshape(self, data):
+        import struct
+        w, h = struct.unpack("<HH", data[6:10])
+        return int(w), int(h)
+        
+
+class JPG(Image):
+
+    imgtype = 'jpg'
+    
+    def _imgshape(self, data):
```

Comment:
I'd make these classmethods and then add simple tests by simply embedding some bytestrings for 1x1 pixel images.

Reply:
Ok, done.  Ready for review/merge.

Reply:
(I used a 2x1 image, as it's testing w,h and it's important not to get those two swapped.)